src/bounds.py

```diff
@@ -5,18 +5,60 @@
 a list of features, here we only concern ourselves with lists of integers.
 """
 
-
-def lower_bound(x: list[int], v: int) -> int:
+# Use Binary Search to compute. 
+def lower_bound(x: list[int], v: int, low=0, high=None) -> int:
     """Get the index of the lower bound of v in x.
 
     If all values in x are smaller than v, return len(x).
+    >>> lower_bound([2,3,5,5,7], 1)
+    0
+    >>> lower_bound([2,3,5,5,7], 2)
+    0
+    >>> lower_bound([2,3,5,5,7], 5)
+    2
+    >>> lower_bound([2,3,5,5,7], 8)
+    5
     """
-    return 0  # FIXME: Obviously the answer isn't always 0
+    high = high or len(x)
+    if low >= high: return len(x) # if all elements in x are smaller than v. 
+    mid = (low+high)//2
+    if x[mid] >= v and x[mid-1] < v:
+        return mid
+    elif x[mid] > v:
+        return lower_bound(x, v, mid+1, high)
+    else: # x[mid] == mid+2
+        return lower_bound(x, v, low, mid)
```

Comment:
You can simplify this a bit, since you don't need to check if `v` falls in `x[mid-1] < v <= x[mid]`.

If you keep moving `low` up to `mid+1` when `v > x[mid]` (and only when `v` is larger than `x[mid]`), then `x[low] <= v` at all times, and it can only become `x[low] == v` when `x[low-1] < v` (because you don't move `low` up to this position unless the previous one (`mid - 1`) was less than `v`. So `low` is always a lower bound, just not the largest, and when `x[low] == v` it must be the lower bound we are going for.

As an invariant, you can say that `x[i] < v` for all `i in range(low)` and `v <= x[i]` for all `i in range(high,n)`. The switch from smaller than `v` to larger or equal happens between `low` and `high`, but if `x[low] == v` then it must necessarily happen *at* `x[low]`.